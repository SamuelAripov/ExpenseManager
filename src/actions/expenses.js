import database from '../firebase/firebase';
import { get } from 'http';

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData
        const expense = { description, note, amount, createdAt }

        return database.ref(`users/${uid}/expenses`).push(expense).then(({ ref }) => {
            console.log(ref)
            dispatch(addExpense({
                id: expense.key,
                ...expense
            }))
        })
    }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
    type: 'REMOVE_EXPENSE',
    id
})
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})
export const clearExpenses = () => ({
    type: 'CLEAR_EXPENSES'
})
export const startSetExpense = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        dispatch(clearExpenses())
        return database
            .ref(`users/${uid}/expenses/`)
            .once('value')
            .then((snapshot) => {
                snapshot.forEach((childSnapshot) => {
                    dispatch(addExpense({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    }))
                    console.log('Official Data', {
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    })
                })
            }).catch(e => console.log(e))
    }
}
export const startRemoveExpense = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(removeExpense({ id }))
            })
    }
}
export const startEditExpense = (id, update) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .update({
                ...update
            })
            .then(() => {
                dispatch(editExpense(id, update))
            })
    }
}

