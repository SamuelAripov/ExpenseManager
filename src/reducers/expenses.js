// Expenses Reducer
import database from '../firebase/firebase'

export default (state = [], action) => {
    switch (action.type) {
        case 'ADD_EXPENSE': {
            console.log(action.expense)
            return [
                ...state,
                action.expense
            ]
        }
        case 'REMOVE_EXPENSE': {
            return state.filter((expense) => expense.id !== action.id)
        }
        case 'EDIT_EXPENSE': {
            return state.map(expense => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else
                    return expense
            })
        }
        case 'SET_EXPENSES': {
            return action.expenses
        }
        case 'CLEAR_EXPENSES': {
            return []
        }
        default:
            return state;
    }
}