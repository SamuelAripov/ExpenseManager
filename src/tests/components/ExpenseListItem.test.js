import React from 'react'
import { shallow } from 'enzyme'
import ListItems from '../../components/ExpenseListItem'
import expenses from '../fixtures/expenses'

test('should render expense list component with expense info', () => {
    const wrapper = shallow(<ListItems {...expenses}/>)
    expect(wrapper).toMatchSnapshot()
})