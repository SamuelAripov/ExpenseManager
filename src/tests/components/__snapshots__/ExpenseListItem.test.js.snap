// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list component with expense info 1`] = `
<div>
  <Link
    to="/edit/undefined"
  >
    <h3>
      Expense | 
    </h3>
  </Link>
  <p>
    Price | 
    $0.00
  </p>
  <p>
    Spent on 
    Jul 23rd 2019
  </p>
  <p />
  <br />
</div>
`;
